game/minecraft_nbt.ksy

```diff
@@ -0,0 +1,225 @@
+
+  This spec can only handle uncompressed NBT data, so be sure to first detect
+  what type of data you are dealing with. You can use the Unix `file` command
+  to do this:
+
+  ```shell
+  file --brief --mime-type input-unknown.nbt
+  ```
+
+  If it says:
+
+    * `application/gzip`, you can decompress it by
+      * `gunzip -c input-gzip.nbt > output.nbt` or
+      * `python3 -c "import sys, gzip; sys.stdout.buffer.write(
+        gzip.decompress(sys.stdin.buffer.read()) )" < input-gzip.nbt > output.nbt`
+    * `application/zlib`, you can use
+      * `openssl zlib -d -in input.nbt.zlib -out output.nbt` (does not work on most systems)
+      * `python3 -c "import sys, zlib; sys.stdout.buffer.write(
+        zlib.decompress(sys.stdin.buffer.read()) )" < input-zlib.nbt > output.nbt`
+    * something else (especially `image/x-pcx` and `application/octet-stream`),
+      it is most likely already uncompressed.
+
+  The file `output.nbt` generated by one of the above commands can already be
+  processed with this Kaitai Struct specification.
+
+  This spec **only** implements the Java edition format. There is also
+  a [Bedrock edition](https://wiki.vg/NBT#Bedrock_edition) NBT format,
+  which uses little-endian encoding and has a few other differences, but it isn't
+  as popular as the Java edition format.
+
+  **Implementation note:** strings in `TAG_String` are incorrectly decoded with
+  standard UTF-8, while they are encoded in [**Modified UTF-8**](
+    https://docs.oracle.com/javase/8/docs/api/java/io/DataInput.html#modified-utf-8
+  ) (MUTF-8). That's because MUTF-8 is usually not supported natively by most target
+  languages, and thus one must use external libraries to achieve a fully-compliant
+  decoder. But decoding in standard UTF-8 is still better than nothing, and
+  it usually works fine.
+
+  All Unicode codepoints with incompatible representations in MUTF-8 and UTF-8 are
+  U+0000 (_NULL_), U+D800-U+DFFF (_High_ and _Low Surrogates_) and U+10000-U+10FFFF
+  (_Supplementary_ Planes; includes e.g. emoticons, pictograms). When a _MUTF-8_-encoded
+  string is decoded as UTF-8, every such code point will be replaced with
+  2 (for _NULL_ character), 4 (for _Surrogates_) or 6 (for _Supplementary_ characters)
+  "�" symbols (i.e. U+FFFD _Replacement Character_).
+
+  ---
+
+  Sample files:
+
+    * <https://wiki.vg/NBT#Download>
+    * <https://github.com/twoolie/NBT/blob/f9e892e/tests/world_test/data/scoreboard.dat>
+    * <https://github.com/chmod222/cNBT/tree/3f74b69/testdata>
+    * <https://github.com/PistonDevelopers/hematite_nbt/tree/0b85f89/tests>
+doc-ref:
+  - https://wiki.vg/NBT
+  - https://web.archive.org/web/20110723210920/https://www.minecraft.net/docs/NBT.txt
+  - https://minecraft.gamepedia.com/NBT_format
+seq:
+  - size: 0
+    if: root_type == tag::end # always `false`, but it doesn't matter
+    doc: Invoke `valid` check before we start parsing
+  - id: root
+    type: named_tag
+instances:
+  root_type:
```

Comment:
I meant

```yaml
instances:
  is_valid:
    value: root.type == tag::compound
    valid: true
```

and it is likely it still won't be compiled.

Reply:
@KOLANICH:

> I meant
> 
> ```yaml
> instances:
>   is_valid:
>     value: root.type == tag::compound
>     valid: true
> ```
> 
> and it is likely it still won't be compiled.

Yes, it does not work on `value` instances.

Reply:
Would be nice, if it is possible to do it.

Reply:
> Why not just put the `valid` attribute directly on the `root` field like this?

This is the first thing which occurred to me, but the problem is that it makes the check execute only when the entire file is parsed, and that is pretty late to realize that all the parsing was unnecessary, because it's apparently a completely different format.